### PR TITLE
PHP: update codegen plugin to support php_namespace option

### LINKED ISSUE #11508: PHP: support php_namespace option in gRPC protoc plugin
PHP Protobuf has recently added support for a php_namespace option (similar to the existing csharp_namespace option) that can be used to specify the namespace of generated message types. See https://github.com/google/protobuf/pull/3162

This option should be supported by the gRPC protoc plugin so that the generated stub is placed in the same namespace as generated protobuf classes, and so that stub methods use the correct types.

cc @stanley-cheung

### PR DESCRIPTION
**Pull request by stanley-cheung**
Fixes #11508

- Support the new "php_namespace" option
- Accept the protoc command line parameter in a more general format `protoc --proto_path=<path> --grpc_out=class_suffix=Foo:<path> --plugin=protoc-gen-grpc=<path>/grpc_php_plugin some.proto` (The param clause can expand out to `--grpc_out=class_suffix=Foo,param2=val2:<path>` in the future)

Need to bump `third_party/protobuf` to the unreleased `3.4.x` branch because the proto "php_namespace" option was introduced in `3.4.x`

cc @michaelbausor @teboring

### CODE REVIEW

File: tools/run_tests/sanity/check_submodules.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ cat << EOF | awk '{ print $1 }' | sort > $want_submodules
  886e7d75368e3f4fab3f4d0d3584e4abfc557755 third_party/boringssl-with-bazel (version_for_cocoapods_7.0-857-g886e7d7)
  30dbc81fb5ffdc98ea9b14b1918bfe4e8779b26e third_party/gflags (v2.2.0)
  ec44c6c1675c25b9827aacd08c02433cccde7780 third_party/googletest (release-1.8.0)
- a6189acd18b00611c1dc7042299ad75486f08a1a third_party/protobuf (v3.3.0)
+ 942a29cecd36f2a4b22fdd2179635cd548e6bd27 third_party/protobuf (3.4.x)
```
Comment (kpayson64) [inline tools/run_tests/sanity/check_submodules.sh]: Any idea when the 3.4.0 release is happening?  I know I got some push back when I tried to move the dependency to a non-release protobuf version.

File: src/compiler/php_plugin.cc [hunk 2]
```diff
@@ -40,12 +41,25 @@ class PHPGrpcGenerator : public grpc::protobuf::compiler::CodeGenerator {
       return true;
     }
 
+    std::vector<std::pair<grpc::string, grpc::string> > options;
+    ParseGeneratorParameter(parameter, &options);
+
+    grpc::string class_suffix;
+    for (size_t i = 0; i < options.size(); ++i) {
+      if (options[i].first == "class_suffix") {
```
Comment (kpayson64) [inline src/compiler/php_plugin.cc]: I'm not sure I understand the motivation for this change.  Wouldn't this break people who were previously specifying the suffix?

File: tools/run_tests/sanity/check_submodules.sh (hunk 1 above)
Comment (stanley-cheung) [inline tools/run_tests/sanity/check_submodules.sh]: The last update I got is end of July. I am OK with holding off until there's a released version (or at least the RC version).

File: src/compiler/php_plugin.cc (hunk 2 above)
Comment (stanley-cheung) [inline src/compiler/php_plugin.cc]: The PR #11773 handling the `parameter` has *just* been merged to `master` yesterday. So no one is using it yet. I want to make changes on top of that to make sure it's more flexible when this does get released.

Comment (stanley-cheung) [inline src/compiler/php_plugin.cc]: C# is also doing something [similar](https://github.com/grpc/grpc/blob/5253c8f9a899450397a5e46e4923d01ac9a66a27/src/compiler/csharp_plugin.cc#L37)

Comment (kpayson64) [inline src/compiler/php_plugin.cc]: If this is a new feature and we have some flexibility, is there any way we can have some kind of protobuf ```grpc_args``` parameter?  It feels really awkward to jam arguments into the ```grpc_out``` parameter.

Might be out of scope for this PR, but just my 2-cents.

Comment (stanley-cheung) [inline src/compiler/php_plugin.cc]: My understanding is that this is the preferred way (by way of how `protoc` works) of passing command line parameters to a protoc plugin. The `parameter` parameter is already there in the `Generate` method in the plugin already supplied by the main `protoc`. So we did not introduce that parameter per se. It's already passed to the plugin if you do specify something before the `:` in `--grpc_out`. This is just to formalize that, if you do pass something before the `:`, it should be in the format of `param1:var1,param2:var2`

File: tools/run_tests/sanity/check_submodules.sh (hunk 1 above)
Comment (stanley-cheung) [inline tools/run_tests/sanity/check_submodules.sh]: Bumped the protobuf version to the `v3.4.0rc1` tag.

**Comment by jtattermusch** [comment]
Sanity failing: 
```
2017-08-08 05:27:06,851 Extracting Bazel installation...
..........................
____Loading package: 
____Loading package: @bazel_tools//tools/cpp
____Loading package: @local_config_cc//
file /var/local/git/grpc/tools/distrib/python/grpcio_tools/protoc_lib_deps.py is out of date

Have you called tools/distrib/python/make_grpcio_tools.py since upgrading protobuf?
```

also, no kokoro tests have run.

**Comment by stanley-cheung** [comment]
Mac failure: #12155 
Windows failure: #11745

**Comment by jtattermusch** [comment]
I think this also broke building the protoc artifact:
https://kokoro.corp.google.com/job/grpc/job/ubuntu/job/master/job/grpc_build_artifacts/400/

**Comment by pherl** [comment]
C++ build was broken because the compiler doesn't support ref-qualifiers even compiled with --std=c++11. I checked the compiler support list:
GCC >= 4.8.1
CLANG >= 2.9
INTEL >= 14.0
MSVC >= Nov 2013 CTP

I think most of them are released in 2013. It's 5 years now, can we update kokoro to a newer compiler?

**Comment by pherl** [comment]
This should be fixed now. Please help verify with the https://github.com/google/protobuf/releases/tag/v3.4.0rc3

**Pull request merged**